### PR TITLE
Spark 3.3: Ensure rowStartPosInBatch in ColumnarBatchReader is set correctly

### PR DESCRIPTION
**Pull request by wypoon**
There is a bug in Parquet vectorized reads reported in https://github.com/apache/iceberg/issues/5927.
This bug happens when reading a Parquet data file (using the `BatchDataReader`) that is bigger than the split size, and there are deletes that need to be applied to the data file. The cause of the bug is that `ColumnarBatchReader#setRowGroupInfo` is not called with the correct `rowPosition`, and that is because in `ReadConf`,  `generateOffsetToStartPos(Schema)` returns null due to an optimization. (When this happens, the `startRowPositions` array is always populated with 0s, and thus `ColumnarBatchReader#setRowGroupInfo` gets called with `rowPosition` 0 even when the `rowPosition` is that of the second or subsequent row group. In `ColumnarBatchReader`, `setRowGroupInfo` initializes a `rowStartPosInBatch`  field, which is used to determine where in the `PositionDeleteIndex` to start applying deletes from. When `rowStartPosInBatch` is incorrectly initialized, the indexes of positional deletes are not correctly aligned with the rows in the data file.)
The fix is to ensure that when there are deletes, the `Schema` has the `_pos` metadata column in it. Then `ReadConf#generateOffsetToStartPos(Schema)` will generate the necessary `Map` that is used to compute the `startRowPositions`.

Added a unit test that reproduces the problem without this fix. Without this fix, the test passes for non-vectorized read and fails for vectorized read. With this fix, the test passes for both cases.

### CODE REVIEW

**Comment by wypoon** [comment]
@flyrain I believe you implemented the support for row-level deletes in the vectorized reader. Can you please review this? Also @aokolnychyi @RussellSpitzer @chenjunjiedada.

**Review by chenjunjiedada — APPROVED** [review]
+1.  Could you please help to backport to spark 3.1 and 3.2 as well?

**Comment by wypoon** [comment]
Thanks for reviewing, @chenjunjiedada. I do plan to port this to Spark 3.2 once this is approved by a committer. Let me look into 3.1 as well. I also think this bug needs to be fixed in the 1.0.x branch.

File: spark/v3.3/build.gradle
Comment (flyrain) [inline spark/v3.3/build.gradle:144]: We don't have to put the test in this module. Basically, a parquet file with multiple row groups will trigger the bug in class `TestSparkReaderDeletes`.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestParquetMergeOnRead.java
Comment (flyrain) [inline spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestParquetMergeOnRead.java:124]: I think the purpose is to test spark row-level delete read in case of multiple row groups. As long as we create multiple row groups in parquet file, we should be good.  We can use the same way in this class to make sure of generating row groups, https://github.com/apache/iceberg/blob/bac84bcc580a392955eb520b9e3c316a16af62b7/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/data/TestSparkParquetReadMetadataColumns.java#L133-L133

**Review by flyrain — COMMENTED** [review]
+1 for the solution. Thanks @wypoon for the fix!
Not a blocker. We may do it in another PR, which I think we need more coverage for multiple parquet row groups in terms of deletes read. We can add an option [here](https://github.com/apache/iceberg/blob/master/data/src/test/java/org/apache/iceberg/data/DeleteReadTests.java#L95-L95) to generate multiple-row-group parquet files, and reuse the class `TestSparkReaderDeletes`.

I'm a bit confused of this behavior: `ReadConf.startRowPositions` is valid only if `_pos` column exists in the `expectedSchema` due to #1716. Are there use cases that `_pos` is absent and we still need `ReadConf.startRowPositions`? By looking at the class `VectorizedParquetReader` and `ParquetReader` who are consuming `ReadConf.startRowPositions`, it seems likely the schema doesn't have `_pos`.
 cc @chenjunjiedada @aokolnychyi

**Comment by wypoon** [comment]
> I'm a bit confused of this behavior: `ReadConf.startRowPositions` is valid only if `_pos` column exists in the `expectedSchema` due to #1716. Are there use cases that `_pos` is absent and we still need `ReadConf.startRowPositions`? By looking at the class `VectorizedParquetReader` and `ParquetReader` who are consuming `ReadConf.startRowPositions`, it seems likely the schema doesn't have `_pos`.

I too was surprised by the behavior. In my example, **before my fix**, when the query
```
select count(*) from default.test_iceberg where e is null
```
is run after the update, the `Schema` that is passed to `ReadConf#generateOffsetToStartPos(Schema)`  is
```
{
  5: e: optional double
}
```
so it did not have `_pos`.

@flyrain are you asking if there are **other** cases where `_pos` will still be absent **after** this fix and we need `ReadConf#startRowPositions()` to return a valid `startRowPositions`?

File: spark/v3.3/build.gradle
Comment (wypoon) [reply spark/v3.3/build.gradle:144]: Let me look into your test pointers.
I put my repro in spark-extensions because I was simply following the example given in https://github.com/apache/iceberg/issues/5927. And in order to use UPDATE, we need the spark-extensions JAR to be present.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestParquetMergeOnRead.java
Comment (wypoon) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestParquetMergeOnRead.java:124]: I agree that the bug simply depends on having multiple row groups in the data file that has applicable positional deletes.

**Comment by chenjunjiedada** [comment]
> 
> I'm a bit confused of this behavior: `ReadConf.startRowPositions` is valid only if `_pos` column exists in the `expectedSchema` due to #1716. Are there use cases that `_pos` is absent and we still need `ReadConf.startRowPositions`? By looking at the class `VectorizedParquetReader` and `ParquetReader` who are consuming `ReadConf.startRowPositions`, it seems likely the schema doesn't have `_pos`. cc @chenjunjiedada @aokolnychyi

The row group start positions are always computed but are only correct when it is projected right now.  That's intended because we don't want to read the parquet footer one more time.  But since the footer must be read at least once, we should be able to cache some content during the first access to avoid the current optimization logic and thus simply the logic to check `_pos` column.

**Comment by flyrain** [comment]
> @flyrain are you asking if there are other cases where _pos will still be absent after this fix and we need ReadConf#startRowPositions() to return a valid startRowPositions?

Yes, I have concerns for this case. But I guess it is fine since `VectorizedArrowReader::setRowGroupInfo()` doesn't consume the `rowPosition`, which means it still works fine even `rowPosition` is off. The problem happens only if we want to rely on the row positions. https://github.com/apache/iceberg/blob/5688d59f6b8ea4784de1f45c9386e13618803673/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java#L396

By digging a bit more, `ReadConf.columnChunkMetadataForRowGroups` keeps the row group metadata. Wondering if we can calculate the row positions while generating the `columnChunkMetadataForRowGroups`, so that the row position will always be correct and we don't have to read the metadata twice. What do you think, @chenjunjiedada? I guess that's what you mentioned to avoid optimization. https://github.com/apache/iceberg/blob/5688d59f6b8ea4784de1f45c9386e13618803673/parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java#L237-L237

**Comment by chenjunjiedada** [comment]
@flyrain, Correct, that is it.

**Comment by wypoon** [comment]
@flyrain for the test part, I followed your suggestion and added a test in `TestSparkReaderDeletes` instead (removing the earlier one).

File: spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java
Comment (wypoon) [inline spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkReaderDeletes.java:599]: Without the fix, this assertion fails for the vectorized case.
There are 3 deletes applied to the first row group and 4 deletes applied to the second row group. Without the fix, the 3 deletes for the first row group are applied to the second as well. Thus 6 rows are deleted and the result is 194 rows, instead of the expected 193.

File: spark/v3.3/build.gradle
Comment (wypoon) [reply spark/v3.3/build.gradle]: Added a test in `TestSparkReaderDeletes` as you suggested, so there is no need to modify the spark-extensions build.

File: spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestParquetMergeOnRead.java
Comment (wypoon) [reply spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestParquetMergeOnRead.java]: Thanks for the pointer.

**Review by flyrain — APPROVED** [review]
+1 Thanks @wypoon for the fix.

**Comment by flyrain** [comment]
> @flyrain, Correct, that is it.

Hi @chenjunjiedada , filed PR https://github.com/apache/iceberg/pull/6056 based on the discussion. Please take a look. Thanks.

**Pull request merged**
